README.md
# modi2-network-nvs-reset-tool

release_notes.md
## Feature
1. Motor pnp 이미지 변경


bootstrap.py
import os

from shutil import rmtree
from platform import system
from argparse import ArgumentParser


def make_clean():
    cwd = os.path.dirname(__file__)
    dirnames = ['__pycache__', 'build', 'dist']
    for d in dirnames:
        dirpath = os.path.join(cwd, d)
        if os.path.isdir(dirpath):
            rmtree(dirpath)

    redundant_specfile = os.path.join(cwd, 'main.spec')
    if os.path.exists(redundant_specfile):
        os.remove(redundant_specfile)

    redundant_logfile = os.path.join(cwd, 'gmfu.log')
    if os.path.exists(redundant_logfile):
        os.remove(redundant_logfile)


def make_executable():
    make_clean()
    os.system('pyinstaller modi+_network_reset.spec')


if __name__ == '__main__':
    parser = ArgumentParser()
    parser.add_argument(
        '--mode', type=str, default='install',
        choices=['clean', 'install'],
        help='This is a script which makes your life a lot easier :)'
    )
    args = parser.parse_args()
    mode = args.mode
    mode_func = {
        'clean': make_clean,
        'install': make_executable,
    }.get(mode)
    mode_func()


modi2_network_nvs_reset/core/network_reset.py
import json
import sys
import threading as th
import time
from base64 import b64decode, b64encode
from io import open
from os import path
import threading
import os

import serial
# import serial.tools.list_ports as stl
import serial.tools.list_ports as sp

from modi2_network_nvs_reset.util.connection_util import SerTask
from modi2_network_nvs_reset.util.message_util import (decode_message,
                                                     parse_message,
                                                     unpack_data)
from modi2_network_nvs_reset.util.module_util import (Module,
                                                    get_module_type_from_uuid)
from modi2_network_nvs_reset.util.module_util import (Module,
                                                    get_module_uuid_from_type)

def retry(exception_to_catch):
    def decorator(func):
        def wrapper(*args, **kwargs):
            try:
                return func(*args, **kwargs)
            except exception_to_catch:
                return wrapper(*args, **kwargs)
        return wrapper
    return decorator

class Network_reset_manager:
    """Module Firmware Updater: Updates a firmware of given module"""
    SERIAL_MODE_COMPORT = 1
    SERIAL_MODI_WINUSB = 2
    
    NO_ERROR = 0
    UPDATE_READY = 1
    WRITE_FAIL = 2
    VERIFY_FAIL = 3
    CRC_ERROR = 4
    CRC_COMPLETE = 5
    ERASE_ERROR = 6
    ERASE_COMPLETE = 7

    UPDATE_FIRMWARE_MODE = 0
    CHNAGE_TYPE_MODE = 1

    FILE_UPLOADE_STATE = 0x15
    FILE_UPLOADE_COMMAND = 0x16

# self.signal_list[2].emit(100," ")
    def __init__(
        self, port=None, conn_type="ser"
    ):
        self.print = True
        self.conn_type = conn_type

        self.app_version = 8192

        self.line = [] #라인 단위로 데이터 가져올 리스트 변수
        self.serial_port = None
        self.baud = 115200
        self.exitThread = False   # 쓰레드 종료용 변수
        self.file_name = ""
        self.start_flag = False
        self.signal_list = []
        self.target_module_health_flag = False
        self.target_uuid = 0
        self.target_id = 0
        self.nvs_reset_timeout_thread = None
        
    def __del__(self):
        try:
            self.close()
        except serial.SerialException:
            self.__print("Magic del is called with an exception")

    def parse_message(self, command: int, source: int, destination: int,
                    byte_data: bytes):
        message = dict()
        message['c'] = command
        message['s'] = source
        message['d'] = destination
        message['b'] = self.__encode_bytes(byte_data)
        message['l'] = len(byte_data)
        return json.dumps(message, separators=(",", ":"))

    def __encode_bytes(self, byte_data: bytes):
        idx = 0
        data = bytearray(len(byte_data))
        while idx < len(byte_data):
            if not byte_data[idx]:
                idx += 1
            elif byte_data[idx] > 256:
                length = self.__extract_length(idx, byte_data)
                data[idx: idx + length] = int.to_bytes(
                    byte_data[idx], byteorder='little', length=length, signed=True
                )
                idx += length
            elif byte_data[idx] < 0:
                data[idx: idx + 4] = int.to_bytes(
                    int(byte_data[idx]), byteorder='little', length=4, signed=True
                )
                idx += 4
            elif byte_data[idx] < 256:
                data[idx] = int(byte_data[idx])
                idx += 1
        return b64encode(bytes(data)).decode('utf8')

    def __extract_length(self, begin: int, src: bytes) -> int:
        length = 1
        for i in range(begin + 1, len(src)):
            if not src[i]:
                length += 1
            else:
                break
        return length

    def uuid_to_module_type(self, uuid):
        module_type_temp = (uuid >> 32) & 0xFFFF
        if module_type_temp == 0x0000:
            return "Network"
        elif module_type_temp == 0x0010:
            return "Battery"
        elif module_type_temp == 0x2000:
            return "Env"
        elif module_type_temp == 0x2010:
            return "Gyro"
        elif module_type_temp == 0x2030:
            return "Button"
        elif module_type_temp == 0x2040:
            return "Dial"
        elif module_type_temp == 0x2070:
            return "Joystick"
        elif module_type_temp == 0x2080:
            return "ToF"
        elif module_type_temp == 0x4000:
            return "Display"
        elif module_type_temp == 0x4010:
            return "MotorA"
        elif module_type_temp == 0x4011:
            return "MotorB"
        elif module_type_temp == 0x4020:
            return "Led"
        elif module_type_temp == 0x4030:
            return "Speaker"
        return "none"

    def type_to_module_uuid(self, uuid):
        if uuid == "Network":
            return 0x0000
        elif uuid == "Battery":
            return 0x0010
        elif uuid ==  "Env":
            return 0x2000
        elif uuid == "Gyro":
            return 0x2010
        elif uuid == "Button":
            return 0x2030
        elif uuid == "Dial":
            return 0x2040
        elif uuid == "Joystick":
            return 0x2070
        elif uuid == "ToF":
            return 0x2080
        elif uuid == "Display":
            return 0x4000
        elif uuid == "MotorA":
            return 0x4010
        elif uuid == "MotorB":
            return 0x4011
        elif uuid == "Led":
            return 0x4020
        elif uuid == "Speaker":
            return 0x4030
        return 1

    def version_to_int(self, version):
        version_value = version[1:].split(".")
        return int(version_value[0]) << 13 | int(version_value[1]) << 8 | int(version_value[2])
    
    def parsing_data(self, data, ser):
        tmp = ''.join(data)
        
        jsonObject = json.loads(tmp)

        cmd = jsonObject.get("c")
        sid = jsonObject.get("s")
        did = jsonObject.get("d")
        length = jsonObject.get("l")
        data = b64decode(jsonObject.get("b"))

        # if cmd != 0 :
        # print(tmp)
        
        if cmd == 0x0:  # health
            if self.start_flag == False:
                send_data = int.to_bytes(0xFFF, byteorder="little", length=8)
                ser.write(parse_message(0x28, 0x0, sid, send_data).encode('utf-8'))
        elif cmd == 0x05: # assign id
            if self.start_flag == False:
                received_uuid = int.from_bytes(data, byteorder='little') & 0xFFFFFFFFFFFF
                print(self.uuid_to_module_type(received_uuid), " module detected, uuid = ", hex(received_uuid))
                if self.uuid_to_module_type(received_uuid) == "Network":
                    self.start_flag = True
                    self.target_uuid = received_uuid
                    self.target_id = self.target_uuid & 0xFFF
                    send_data = int.to_bytes(0, byteorder="little", length=1)
                    ser.write(parse_message(0x04, 30, self.target_id, send_data).encode('utf-8')) # Reset
                    self.signal_list[1].emit("module detected")
                    self.nvs_reset_timeout_thread.start()
        elif cmd == 0x0A:
            print("warning")
        elif cmd == 0xA1:
            if self.start_flag == True:
                if sid == 9: # esp32 version
                    self.signal_list[1].emit("reset complete\npress the button")
                    self.signal_list[2].emit()
                    self.start_flag = False
                    ser.close()
                    self.exitThread = True


    def readThread(self, ser):
        json_flag = False
        while not self.exitThread:
            for c in ser.read(1):
                if json_flag == False:
                    if chr(c) == '{':
                        self.line.append(chr(c))
                        json_flag = True
                else :
                    self.line.append(chr(c))
                    if chr(c) == '}':
                        try:
                            json_flag = False
                            self.parsing_data(self.line, ser)
                        except Exception as e:
                            del self.line[:]
                            error_string = repr(e)
                            print(error_string)
                            if "The version of this module" in error_string:
                                raise e 
                            elif "Streaming response error" in error_string:
                                raise e
                        del  self.line[:]

    def nvs_reset_timeout_thread_function(self, ser):
        timeout_count = 0
        while not self.exitThread:
            time.sleep(2)
            if timeout_count < 3 and self.start_flag == True:
                send_data = int.to_bytes(0, byteorder="little", length=1)
                ser.write(parse_message(0x04, 30, self.target_id, send_data).encode('utf-8')) # Reset
                timeout_count += 1
            else:
                if self.start_flag == False:
                    return
                print("Timeout error")
                self.signal_list[0].emit("Timeout error")
                self.signal_list[1].emit("Press the button")
                self.signal_list[2].emit()
                self.exitThread = True
                ser.close()

    def start_reset_thread(self):
        ports = sp.comports()
        ser = None

        for port in ports:
            if (port.vid == 0x2FDE and port.pid == 0x0003):
                self.serial_port = port.device

        if self.serial_port is None:
            if sys.platform.startswith("win"):
                from modi2_network_nvs_reset.util.modi_winusb.modi_winusb import list_modi_winusb_paths
                path_list = list_modi_winusb_paths()
                for index, value in enumerate(path_list):
                    self.serial_port = value
        else:
            self.type = self.SERIAL_MODE_COMPORT
            ser = serial.Serial(self.serial_port, self.baud, timeout=0)

        if self.serial_port is None:
            print("Please connect MODI+ Network Module")
            self.signal_list[0].emit("Please connect MODI+ Network Module")
            self.signal_list[1].emit("Press the button")
            self.signal_list[2].emit()
            return
        else:
            if sys.platform.startswith("win"):
                from modi2_network_nvs_reset.util.modi_winusb.modi_winusb import ModiWinUsbComPort, list_modi_winusb_paths
                if self.serial_port in list_modi_winusb_paths():
                    self.type = self.SERIAL_MODI_WINUSB
                    winusb = ModiWinUsbComPort(path = self.serial_port, baudrate=self.baud, timeout=0)
                    ser = winusb

        self.nvs_reset_timeout_thread = threading.Thread(target=self.nvs_reset_timeout_thread_function, args=(ser,), daemon=True)
        thread1 = threading.Thread(target=self.readThread, args=(ser,), daemon=True)
        thread1.start()

    def set_ui(self, ui):
        self.ui = ui
        print(type(ui))
    
    def set_ui(self, ui, signal_list):
        self.ui = ui
        self.signal_list = signal_list

    def set_print(self, print):
        self.print = print

    def set_raise_error(self, raise_error_message):
        self.raise_error_message = raise_error_message

    def request_network_id(self):
        self.__conn.send_nowait(
            parse_message(0x28, 0x0, 0xFFF, (0xFF, 0x0F))
        )

    def close(self):
        self.__running = False

    def add_to_waitlist(self, module_id: int, module_type: str) -> None:
        # Check if input module already exist in the list
        for curr_module_id, curr_module_type in self.modules_to_update:
            if module_id == curr_module_id:
                return

        # Check if module is already updated
        for curr_module_id, curr_module_type in self.modules_updated:
            if module_id == curr_module_id:
                return

        self.__print(
            f"Adding {module_type} ({module_id}) to waiting list..."
            f"{' ' * 60}"
        )

        # Add the module to the waiting list
        module_elem = module_id, module_type
        self.modules_to_update.append(module_elem)
        self.update_module_num += 1

    @staticmethod
    def __delay(span):
        init_time = time.perf_counter()
        while time.perf_counter() - init_time < span:
            pass
        return

    @staticmethod
    def __set_module_state(
        destination_id: int, module_state: int, pnp_state: int
    ) -> str:
        message = dict()

        message["c"] = 0x09
        message["s"] = 0
        message["d"] = destination_id

        state_bytes = bytearray(2)
        state_bytes[0] = module_state
        state_bytes[1] = pnp_state

        message["b"] = b64encode(bytes(state_bytes)).decode("utf-8")
        message["l"] = 2

        return json.dumps(message, separators=(",", ":"))

    def __print(self, data, end="\n"):
        if self.print:
            print(data, end)

modi2_network_nvs_reset/gui_firmware_upload.py
import logging
import os
import pathlib
import sys
import threading as th
import time
import traceback as tb
import io
import urllib.request as ur
import zipfile
import shutil
from io import open
from os import path
from urllib.error import URLError

from PyQt5 import QtCore, QtGui, QtWidgets, uic
from PyQt5.QtCore import QObject, pyqtSignal, pyqtSlot
from PyQt5.QtWidgets import QDialog, QFileDialog

from modi2_network_nvs_reset.util.connection_util import list_modi_ports
# from modi2_network_nvs_reset.core.network_uploader import NetworkFirmwareMultiUpdater
from modi2_network_nvs_reset.core.network_reset import Network_reset_manager


class StdoutRedirect(QObject):
    printOccur = pyqtSignal(str, str, name="print")

    def __init__(self):
        QObject.__init__(self, None)
        self.daemon = True
        self.sysstdout = sys.stdout.write
        self.sysstderr = sys.stderr.write
        self.logger = None

    def stop(self):
        sys.stdout.write = self.sysstdout
        sys.stderr.write = self.sysstderr

    def start(self):
        sys.stdout.write = self.write
        sys.stderr.write = lambda msg: self.write(msg, color="red")

    def write(self, s, color="black"):
        sys.stdout.flush()
        self.printOccur.emit(s, color)
        if self.logger and not self.__is_redundant_line(s):
            self.logger.info(s)

    @staticmethod
    def __is_redundant_line(line):
        return (
            line.startswith("\rUpdating") or
            line.startswith("\rFirmware Upload: [") or
            len(line) < 3
        )


class PopupMessageBox(QtWidgets.QMessageBox):
    def __init__(self, main_window, level):
        QtWidgets.QMessageBox.__init__(self)
        self.window = main_window
        self.setSizeGripEnabled(True)
        self.setWindowTitle("System Message")

        def error_popup():
            self.setIcon(self.Icon.Warning)
            self.setText("ERROR")

        def warning_popup():
            self.setIcon(self.Icon.Information)
            self.setText("WARNING")
            self.addButton("Ok", self.ActionRole)
            # restart_btn.clicked.connect(self.restart_btn)

        func = {
            "error": error_popup,
            "warning": warning_popup,
        }.get(level)
        func()

        close_btn = self.addButton("Exit", self.ActionRole)
        close_btn.clicked.connect(self.close_btn)
        # report_btn = self.addButton('Report Error', self.ActionRole)
        # report_btn.clicked.connect(self.report_btn)
        self.show()

    def event(self, e):
        MAXSIZE = 16_777_215
        MINHEIGHT = 100
        MINWIDTH = 200
        MINWIDTH_CHANGE = 500
        result = QtWidgets.QMessageBox.event(self, e)

        self.setMinimumHeight(MINHEIGHT)
        self.setMaximumHeight(MAXSIZE)
        self.setMinimumWidth(MINWIDTH)
        self.setMaximumWidth(MAXSIZE)
        self.setSizePolicy(
            QtWidgets.QSizePolicy.Expanding, QtWidgets.QSizePolicy.Expanding
        )

        textEdit = self.findChild(QtWidgets.QTextEdit)
        if textEdit is not None:
            textEdit.setMinimumHeight(MINHEIGHT)
            textEdit.setMaximumHeight(MAXSIZE)
            textEdit.setMinimumWidth(MINWIDTH_CHANGE)
            textEdit.setMaximumWidth(MAXSIZE)
            textEdit.setSizePolicy(
                QtWidgets.QSizePolicy.Expanding,
                QtWidgets.QSizePolicy.Expanding,
            )

        return result

    def close_btn(self):
        self.window.close()

    def report_btn(self):
        pass

class ThreadSignal(QObject):
    thread_error = pyqtSignal(object)
    thread_signal = pyqtSignal(object)

    def __init__(self):
        super().__init__()


class Form(QDialog):
    """
    GUI Form of MODI Firmware Updater
    """
    current_module_changed_signal = pyqtSignal(int, str)
    progress_signal = pyqtSignal(int)
    button_enable_signal = pyqtSignal()
    error_message_signal = pyqtSignal(str)
    button_text_change_signal = pyqtSignal(str)

    def __init__(self, installer=False):
        QDialog.__init__(self)
        # self.logger = self.__init_logger()
        self.__excepthook = sys.excepthook
        sys.excepthook = self.__popup_excepthook
        th.excepthook = self.__popup_thread_excepthook
        self.err_list = list()
        self.is_popup = False

        # ui_path = os.path.join(os.path.dirname(__file__), "assets", "file-uploader.ui")
        ui_path = os.path.join(os.path.dirname(__file__), "assets", "network-reset.ui")

        if sys.platform.startswith("win"):
            self.component_path = pathlib.PurePosixPath(pathlib.PurePath(__file__), "..", "assets", "component")
        else:
            self.component_path = os.path.join(os.path.dirname(__file__), "assets", "component")

        self.test_signal_list = []

        self.error_message_signal.connect(self.set_error_message)
        self.button_text_change_signal.connect(self.set_process_state_text)
        self.button_enable_signal.connect(self.set_button_enable)

        self.test_signal_list.append(self.error_message_signal)
        self.test_signal_list.append(self.button_text_change_signal)
        self.test_signal_list.append(self.button_enable_signal)

        
        version_path = os.path.join(os.path.dirname(__file__), "..", "version.txt")
        with io.open(version_path, "r") as version_file:
            self.version_info = version_file.readline().rstrip("\n")

        self.ui = uic.loadUi(ui_path)
        self.ui.setWindowTitle("MODI+ Network reset - " + self.version_info)
        self.ui.setWindowIcon(QtGui.QIcon(os.path.join(self.component_path, "network_module.ico")))

        self.ui.setStyleSheet("background-color: white")
        # Set signal for thread communication
        self.stream = ThreadSignal()

        # Connect up the buttons
        self.ui.nvs_reset_start.clicked.connect(self.reset_network_module)

        self.buttons = [
            self.ui.nvs_reset_start,
        ]
        # Print init status
        time_now_str = time.strftime("[%Y/%m/%d@%X]", time.localtime())
        print(time_now_str + " GUI MODI Network reset has been started!")

        # Set up field variables
        self.firmware_updater = None
        self.button_in_english = False
        self.console = False

        self.specific_file_path = None

        self.ui.stream = self.stream
        self.ui.popup = self._thread_signal_hook

        self.ui.show()

    def set_error_message(self, error_message):
        QtWidgets.QMessageBox.warning(self, 'error', error_message)

    def set_process_state_text(self, button_text):
        self.ui.process_state.setText(button_text)

    def set_button_enable(self):
        self.ui.nvs_reset_start.setEnabled(True)


    #
    # Main methods
    #
    def reset_network_module(self):
        print("button clicked")
        self.ui.process_state.setText("processing")
        self.ui.nvs_reset_start.setEnabled(False)
        nvs_reset_manager = Network_reset_manager()
        nvs_reset_manager.set_ui(self.ui, self.test_signal_list)
        th.Thread(
            target=nvs_reset_manager.start_reset_thread,
            daemon=True,
        ).start()

    def specific_file_button_event(self):
        fname = QFileDialog.getOpenFileNames(self, 'Open files', './')
        print(fname)
        self.specific_file_path = (fname[0])
        # print(type(self.specific_file_path))
        # print(len(self.specific_file_path))
        self.ui.specific_file_name.setText(str(fname))

    def specific_file_check_box_event(self):
        if self.ui.specific_file_upload_check_box.isChecked():
            self.ui.specific_file_open_button.setEnabled(True)
        else :
            self.ui.specific_file_open_button.setEnabled(False)

    def check_module_firmware(self):
        if not os.path.exists(self.local_firmware_path):
            os.mkdir(self.local_firmware_path)

        self.__check_module_version()
        self.__check_network_base_version()
        self.__check_esp32_version()

    def __check_module_version(self):
        if not os.path.exists(self.local_module_firmware_path):
            assert_path = path.join(path.dirname(__file__), "assets", "firmware", "module")
            shutil.copytree(assert_path, self.local_module_firmware_path)

    def __check_network_base_version(self):
        if not os.path.exists(self.local_network_firmware_path):
            assert_path = path.join(path.dirname(__file__), "assets", "firmware", "module")
            shutil.copytree(assert_path, self.local_network_firmware_path)

    def __check_esp32_version(self):
        if not os.path.exists(self.local_esp32_firmware_path):
            assert_path = path.join(path.dirname(__file__), "assets", "firmware", "esp32")
            shutil.copytree(assert_path, self.local_esp32_firmware_path)
    #
    # Helper functions
    #
    @staticmethod
    def __init_logger():
        logger = logging.getLogger("GUI MODI Firmware Updater Logger")
        logger.setLevel(logging.DEBUG)

        formatter = logging.Formatter("%(asctime)s - %(name)s - %(levelname)s - %(message)s")
        file_handler = logging.FileHandler("gmfu.log")
        file_handler.setLevel(logging.DEBUG)
        file_handler.setFormatter(formatter)

        return logger

    def __popup_excepthook(self, exctype, value, traceback):
        self.__excepthook(exctype, value, traceback)
        if self.is_popup:
            return
        self.popup = PopupMessageBox(self.ui, level="error")
        self.popup.setInformativeText(str(value))
        self.popup.setDetailedText(str(tb.extract_tb(traceback)))
        self.is_popup = True

    def __popup_thread_excepthook(self, err_msg):
        if err_msg.exc_type in self.err_list:
            return
        self.err_list.append(err_msg.exc_type)
        self.stream.thread_error.connect(self.__thread_error_hook)
        self.stream.thread_error.emit(err_msg)

    @pyqtSlot(object)
    def __thread_error_hook(self, err_msg):
        self.__popup_excepthook(
            err_msg.exc_type, err_msg.exc_value, err_msg.exc_traceback
        )

    @pyqtSlot(object)
    def _thread_signal_hook(self):
        self.thread_popup = PopupMessageBox(self.ui, level="warning")
        if self.button_in_english:
            text = (
                "Reconnect network module and "
                "click the button again please."
            )
        else:
            text = "네트워크 모듈을 재연결 후 버튼을 다시 눌러주십시오."
        self.thread_popup.setInformativeText(text)
        self.is_popup = True

    def __click_motion(self, button_type, start_time):
        # Busy wait for 0.2 seconds
        while time.time() - start_time < 0.2:
            pass

        if button_type in [6, 7]:
            self.buttons[button_type].setStyleSheet(f"border-image: url({self.language_frame_path}); font-size: 13px")
        else:
            self.buttons[button_type].setStyleSheet(f"border-image: url({self.active_path}); font-size: 16px")
            for i, q_button in enumerate(self.buttons):
                if i in [button_type, 6, 7]:
                    continue
                q_button.setStyleSheet(f"border-image: url({self.inactive_path}); font-size: 16px")
                q_button.setEnabled(False)

    def __append_text_line(self, line):
        self.ui.console.moveCursor(
            QtGui.QTextCursor.End, QtGui.QTextCursor.MoveAnchor
        )
        self.ui.console.moveCursor(
            QtGui.QTextCursor.StartOfLine, QtGui.QTextCursor.MoveAnchor
        )
        self.ui.console.moveCursor(
            QtGui.QTextCursor.End, QtGui.QTextCursor.KeepAnchor
        )

        # Remove new line character if current line represents update_progress
        if self.__is_update_progress_line(line):
            self.ui.console.textCursor().removeSelectedText()
            self.ui.console.textCursor().deletePreviousChar()

        # Display user text input
        self.ui.console.moveCursor(QtGui.QTextCursor.End)
        self.ui.console.insertPlainText(line)
        # QtWidgets.QApplication.processEvents(
        #     QtCore.QEventLoop.ExcludeUserInputEvents
        # )

    @staticmethod
    def __is_update_progress_line(line):
        return line.startswith("\rUpdating") or line.startswith("\rFirmware Upload: [")

modi2_network_nvs_reset/util/module_util.py
import time
from os import path
from typing import Union

from modi2_network_nvs_reset.util.message_util import parse_message

BROADCAST_ID = 0xFFF


def get_module_type_from_uuid(uuid):
    hexadecimal = hex(uuid).lstrip("0x")
    hexadecimal = int(hexadecimal, 16) >> 32
    type_indicator = hexadecimal
    module_type = {
        # Setup modules
        0x10  : "battery",
        # Input modules
        0x2000: "env",
        0x2010: "gyro",
        0x2020: "mic",
        0x2030: "button",
        0x2040: "dial",
        0x2050: "ultrasonic",
        0x2060: "ir",
        0x2070: "joystick",
        0x2080: "tof",
        # Output modules
        0x4000: "display",
        0x4010: "motor",
        0x4011: "motor",
        0x4020: "led",
        0x4030: "speaker",
    }.get(type_indicator)
    return "network" if module_type is None else module_type

def get_module_uuid_from_type(module_type):
    module_type_hex = {
        # Setup modules
        "battery"   :0x10  ,
        # module :# Input
        "env"       :0x2000,
        "gyro"      :0x2010,
        "mic"       :0x2020,
        "button"    :0x2030,
        "dial"      :0x2040,
        "ultrasonic" :0x2050,
        "ir"        :0x2060,
        "joystick"  :0x2070,
        "tof"       :0x2080,
        #  module :# Output
        "display"   :0x4000,
        "motor_a"   :0x4010,
        "motor_b"   :0x4011,
        "led"       :0x4020,
        "speaker"   :0x4030,
    }.get(module_type)
    return "network" if module_type_hex is None else module_type_hex


class Module:
    """
    :param int id_: The id of the module.
    :param int uuid: The uuid of the module.
    """

    class Property:
        def __init__(self, value: Union[int, float] = 0):
            self.value = value
            self.last_update_time = time.time()

    RUN = 0
    WARNING = 1
    FORCED_PAUSE = 2
    ERROR_STOP = 3
    UPDATE_FIRMWARE = 4
    UPDATE_FIRMWARE_READY = 5
    REBOOT = 6
    PNP_ON = 1
    PNP_OFF = 2

    def __init__(self, id_, uuid, conn_task):
        self._id = id_
        self._uuid = uuid
        self._conn = conn_task

        self.module_type = str()
        self._properties = dict()
        self._topology = {"r": 0, "t": 0, "l": 0, "b": 0}

        # sampling_rate = (100 - property_sampling_frequency) * 11, in ms
        self.prop_samp_freq = 91

        self.is_connected = True
        self.has_printed = False
        self.last_updated = time.time()
        self.battery = 100
        self.position = (0, 0)
        self.__version = None
        self.user_code_status = -1  # 1 if user code and 0 if not

    def __gt__(self, other):
        if self.order == other.order:
            if self.position[0] == other.position[0]:
                return self.position[1] < other.position[1]
            else:
                return self.position[0] > other.position[0]
        else:
            return self.order > other.order

    def __lt__(self, other):
        if self.order == other.order:
            if self.position[0] == other.position[0]:
                return self.position[1] > other.position[1]
            else:
                return self.position[0] < other.position[0]
        else:
            return self.order < other.order

    def __str__(self):
        return f"{self.__class__.__name__} ({self._id})"

    @property
    def has_user_code(self):
        return self.user_code_status == 1

    @property
    def version(self):
        version_string = ""
        version_string += str(self.__version >> 13) + "."
        version_string += str(self.__version % (2 ** 13) >> 8) + "."
        version_string += str(self.__version % (2 ** 8))
        return version_string

    @version.setter
    def version(self, version_info):
        self.__version = version_info

    @property
    def order(self):
        return self.position[0] ** 2 + self.position[1] ** 2

    @property
    def id(self) -> int:
        return self._id

    @property
    def uuid(self) -> int:
        return self._uuid

    @property
    def is_up_to_date(self):
        root_path = path.join(
            path.dirname(__file__), "..", "assets", "firmware", "module"
        )
        version_path = path.join(root_path, "version.txt")
        with open(version_path) as version_file:
            version_info = version_file.readline().lstrip("v").rstrip("\n")
        version_digits = [int(digit) for digit in version_info.split(".")]
        latest_version = (
            version_digits[0] << 13
            | version_digits[1] << 8
            | version_digits[2]
        )
        return latest_version <= self.__version

    def _get_property(self, property_type: int) -> float:
        """Get module property value and request

        :param property_type: Type of the requested property
        :type property_type: int
        """

        # Register property if not exists
        if property_type not in self._properties:
            self._properties[property_type] = self.Property()
            self.__request_property(self._id, property_type)

        # Request property value if not updated for 1.5 sec
        last_update = self._properties[property_type].last_update_time
        if time.time() - last_update > 1.5:
            self.__request_property(self._id, property_type)

        return self._properties[property_type].value

    def update_property(
        self, property_type: int, property_value: float
    ) -> None:
        """Update property value and time

        :param property_type: Type of the updated property
        :type property_type: int
        :param property_value: Value to update the property
        :type property_value: float
        """
        if property_type not in self._properties:
            self._properties[property_type] = self.Property()
        self._properties[property_type].value = property_value
        self._properties[property_type].last_update_time = time.time()

    def __request_property(
        self, destination_id: int, property_type: int
    ) -> None:
        """Generate message for request property

        :param destination_id: Id of the destination module
        :type destination_id: int
        :param property_type: Type of the requested property
        :type property_type: int
        :return: None
        """
        self._properties[property_type].last_update_time = time.time()
        req_prop_msg = parse_message(
            0x03,
            0,
            destination_id,
            (property_type, None, self.prop_samp_freq, None),
        )
        self._conn.send(req_prop_msg)
